File: push_swap_prototype/lab/libs/push_swap/src/chunk_split.c
#include "models.h"

void	innit_size(t_chunk *min, t_chunk *mid, t_chunk *max);
void	set_split_loc(t_loc loc, t_chunk *min, t_chunk *mid, t_chunk *max);
void	set_third_pivots(t_loc loc, int crt_size, int *pivot_1, int *pivot_2);
int	chunk_max_value(t_ps *data, t_chunk *chunk);
int	chunk_value(t_ps *data, t_chunk *chunk, int n);
int	move_from_to(t_ps *data, t_loc from, t_loc to);
void	split_max_reduction(t_ps *data, t_chunk *max);
bool	a_partly_sort(t_ps *data, int n);
void	easy_sort(t_ps *data, t_chunk *chunk);


void	chunk_split(t_ps *data, t_chunk *to_split, t_split_dest *dest)
{
	int	pivot_1;
	int	pivot_2;
	int	max_value;
	int	next_value;

	innit_size(&dest->min, &dest->mid, &dest->max);
	set_split_loc(to_split->loc, &dest->min, &dest->mid, &dest->max);
	set_third_pivots(to_split->loc, to_split->size, &pivot_1, &pivot_2);
	max_value = chunk_max_value(data, to_split);
	while (to_split->size--)
	{
		next_value = chunk_value(data, to_split, 1);
		if (next_value > max_value - pivot_2)
		{
			dest->max.size += move_from_to(data, to_split->loc, dest->max.loc);
			split_max_reduction(data, &dest->max);
			if (a_partly_sort(data, 1) && to_split->size)
				easy_sort(data, to_split);
		}
		else if (next_value > max_value - pivot_1)
			dest->mid.size += move_from_to(data, to_split->loc, dest->mid.loc);
		else
			dest->min.size += move_from_to(data, to_split->loc, dest->min.loc);
	}
}

void	innit_size(t_chunk *min, t_chunk *mid, t_chunk *max)
{
	min->size = 0;
	mid->size = 0;
	max->size = 0;
}

void	set_split_loc(t_loc loc, t_chunk *min, t_chunk *mid, t_chunk *max)
{
	if (loc == TOP_A)
	{
		min->loc = BOTTOM_B;
		mid->loc = TOP_B;
		max->loc = BOTTOM_A;
	}
	else if (loc == BOTTOM_A)
	{
		min->loc = BOTTOM_B;
		mid->loc = TOP_B;
		max->loc = TOP_A;
	}
	else if (loc == TOP_B)
	{
		min->loc = BOTTOM_B;
		mid->loc = BOTTOM_A;
		max->loc = TOP_A;
	}
	else if (loc == BOTTOM_B)
	{
		min->loc = TOP_B;
		mid->loc = BOTTOM_A;
		max->loc = TOP_A;
	}
}

void	set_third_pivots(t_loc loc, int crt_size, int *pivot_1, int *pivot_2)
{
	*pivot_2 = crt_size / 3;
	*pivot_1 = crt_size / 2; // Default initialization
	if (loc == TOP_A || loc == BOTTOM_A)
		*pivot_1 = 2 * crt_size / 3;
	if (loc == TOP_B || loc == BOTTOM_B)
		*pivot_1 = crt_size / 2;
	if ((loc == TOP_A || loc == BOTTOM_A) && crt_size < 15)
		*pivot_1 = crt_size;
	if (loc == BOTTOM_B && crt_size < 8)
		*pivot_2 = crt_size / 2;
}

File: push_swap_prototype/src/internal/solver/chunk_split.go
package solver

import (
	"fmt"
	"push_swap_prototype/internal/chunk"
	"push_swap_prototype/internal/ops"
)

// Returns SplitDest instead of modifying a pointer parameter
func ChunkSplit(ps *ops.SortingState, to_split *chunk.Chunk, dest *chunk.SplitDest) {
	
	fmt.Printf("=== CHUNK_SPLIT START ===\n")
	fmt.Printf("Splitting chunk: loc=%d, size=%d\n", chunk.GetLoc(to_split), chunk.GetSize(to_split))
	
	
	var pivot1, pivot2 int
	initSize(&dest.Min, &dest.Mid, &dest.Max)
	setSplitLoc(chunk.GetLoc(to_split), &dest.Min, &dest.Mid, &dest.Max)
	setThirdPivots(chunk.GetLoc(to_split), chunk.GetSize(to_split), &pivot1, &pivot2)

	maxValue := ChunkMaxValue(ps, to_split)


	for i := 0; i < chunk.GetSize(to_split); i++ {
		nextValue := ChunkValue(ps, to_split, i) // Use position i, not hardcoded 1

		if nextValue > maxValue-pivot2 {
			chunk.SetSize(&dest.Max, chunk.GetSize(&dest.Max)+MoveFromTo(ps, chunk.GetLoc(to_split), chunk.GetLoc(&dest.Max)))
			SplitMaxReduction(ps, &dest.Max)
			if APartlySort(ps, 1) && chunk.GetSize(to_split)-i-1 > 0 {
				fmt.Printf("[%d] Calling EasySort after SplitMaxReduction\n", i+1)
				EasySort(ps, to_split)
			}
		} else if nextValue > maxValue-pivot1 {
			chunk.SetSize(&dest.Mid, chunk.GetSize(&dest.Mid)+MoveFromTo(ps, chunk.GetLoc(to_split), chunk.GetLoc(&dest.Mid)))
		} else {
			chunk.SetSize(&dest.Min, chunk.GetSize(&dest.Min)+MoveFromTo(ps, chunk.GetLoc(to_split), chunk.GetLoc(&dest.Min)))
		}
	}
}

// initSize initializes the size of all destination chunks to 0
func initSize(min, mid, max *chunk.Chunk) {
	chunk.SetSize(min, 0)
	chunk.SetSize(mid, 0)
	chunk.SetSize(max, 0)
}

// setSplitLoc sets the location for each destination chunk based on the source location
// Exactly like C implementation
func setSplitLoc(loc chunk.Loc, min, mid, max *chunk.Chunk) {
	
	switch loc {
	case chunk.TOP_A:
		chunk.SetLoc(min, chunk.BOTTOM_B)
		chunk.SetLoc(mid, chunk.TOP_B)
		chunk.SetLoc(max, chunk.BOTTOM_A)
	case chunk.BOTTOM_A:
		chunk.SetLoc(min, chunk.BOTTOM_B)
		chunk.SetLoc(mid, chunk.TOP_B)
		chunk.SetLoc(max, chunk.TOP_A)
	case chunk.TOP_B:
		chunk.SetLoc(min, chunk.BOTTOM_B)
		chunk.SetLoc(mid, chunk.BOTTOM_A)
		chunk.SetLoc(max, chunk.TOP_A)
	case chunk.BOTTOM_B:
		chunk.SetLoc(min, chunk.TOP_B)
		chunk.SetLoc(mid, chunk.BOTTOM_A)
		chunk.SetLoc(max, chunk.TOP_A)
	}
}

// setThirdPivots sets the pivot values for splitting based on location and size
func setThirdPivots(loc chunk.Loc, crt_size int, pivot1, pivot2 *int) {
	
	*pivot2 = crt_size / 3
	
	switch loc {
	case chunk.TOP_A, chunk.BOTTOM_A:
		*pivot1 = 2 * crt_size / 3
		// Special case for small chunks (matches C: if (crt_size < 15) *pivot_1 = crt_size)
		if crt_size < 15 {
			*pivot1 = crt_size
		}
	case chunk.TOP_B, chunk.BOTTOM_B:
		*pivot1 = crt_size / 2
		if loc == chunk.BOTTOM_B && crt_size < 8 {
			*pivot2 = crt_size / 2
		}
	default:
		*pivot1 = crt_size / 2
	}
}
